test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -0,0 +1,875 @@
+    ) -> Tuple[FSDP, torch.optim.Optimizer, FSDP, torch.optim.Optimizer]:
+        """
+        Returns a pair of (FSDP model, optimizer) for ``use_orig_params=False``
+        and ``True``, respectively.
+        """
+        LR = 1e-2
+        fsdp_kwargs = {
+            "sharding_strategy": sharding_strategy,
+            "cpu_offload": cpu_offload,
+            "use_orig_params": False,
+        }
+        fsdp_model = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.RECURSIVE,
+            CUDAInitMode.CUDA_BEFORE,
+            fsdp_kwargs=fsdp_kwargs,
+            deterministic=True,
+        )
+        optim = torch.optim.Adam(fsdp_model.parameters(), lr=LR)
+        fsdp_kwargs["use_orig_params"] = True
+        fsdp_model_orig_params = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.RECURSIVE,
+            CUDAInitMode.CUDA_BEFORE,
+            fsdp_kwargs=fsdp_kwargs,
+            deterministic=True,
+        )
+        optim_orig_params = torch.optim.Adam(fsdp_model_orig_params.parameters(), lr=LR)
+        return fsdp_model, optim, fsdp_model_orig_params, optim_orig_params
+
+    def _check_fsdp_parameter_parity(self, fsdp1: FSDP, fsdp2: FSDP) -> None:
+        """Checks that two FSDP instances have the same model parameters."""
+        with FSDP.summon_full_params(fsdp1):
+            with FSDP.summon_full_params(fsdp2):
+                for (n1, p1), (n2, p2) in zip(
+                    fsdp1.named_parameters(),
+                    fsdp2.named_parameters(),
+                ):
+                    self.assertEqual(n1, n2)
+                    torch.testing.assert_close(p1, p2)
+
+    def _get_fsdp_parity_subtest_config(self):
+        return {
+            "sharding_strategy": [
+                ShardingStrategy.NO_SHARD,
+                ShardingStrategy.SHARD_GRAD_OP,
+                ShardingStrategy.FULL_SHARD,
+            ],
+        }
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("offload_params", [False, True])
+    def test_multiple_forward(self, offload_params: bool):
+        """
+        Tests that ``use_orig_params=True`` has parity with ``False`` when
+        running multiple forward passes before a backward pass.
+        """
+        cpu_offload = CPUOffload(offload_params=offload_params)
+        self.run_subtests(
+            self._get_fsdp_parity_subtest_config(),
+            self._test_multiple_forward,
+            cpu_offload=cpu_offload,
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def _test_multiple_forward(
+        self,
+        sharding_strategy: ShardingStrategy,
+        cpu_offload: CPUOffload,
+    ):
+        (
+            fsdp_model,
+            optim,
+            fsdp_model_orig_params,
+            optim_orig_params,
+        ) = self._get_fsdp_models_and_optims(sharding_strategy, cpu_offload)
+        device = torch.device("cuda")
+        for i in range(3):
+            optim.zero_grad()
```

Comment:
do we need to test set_to_none here as well or is that already well covered?

Reply:
It is covered in `test_grad_writeback()`.